### PR TITLE
Feature/coverage

### PR DESCRIPTION
**Pull request by jgebal**

### CODE REVIEW

File: source/core/coverage/ut_coverage_helper.pkb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  function  is_develop_mode return boolean is
+  begin
+    return g_develop_mode;
+  end;
+
+  function coverage_start(a_run_comment varchar2) return integer is
+    pragma autonomous_transaction;
+  begin
+    --those are Global Temporary tables for profiler usage only
+    delete from plsql_profiler_data;
```
Comment (Pazus) [inline source/core/coverage/ut_coverage_helper.pkb]: This is not usually done when you use standard profiler tables as they are not temp

File: source/core/coverage/ut_coverage_helper.pkb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  function get_raw_coverage_data(a_object_owner varchar2, a_object_name varchar2) return unit_line_calls is
+    type coverage_row is record (
+      line  binary_integer,
+      calls number(38,0)
+    );
+    type coverage_rows is table of coverage_row;
+    l_tmp_data coverage_rows;
+    l_results  unit_line_calls;
+  begin
+      select d.line#, d.total_occur
```
Comment (Pazus) [inline source/core/coverage/ut_coverage_helper.pkb]: I think its better to use simple for loop as the collection might be big in real life cases an there is no need to select it all at once as you line-by-line convert it later

File: source/reporters/ut_coverage_json_reporter.tps [hunk 3]
```diff
@@ -0,0 +1,60 @@
+create or replace type ut_coverage_json_reporter force under ut_coverage_reporter_base(
```
Comment (Pazus) [inline source/reporters/ut_coverage_json_reporter.tps]: lets not use `force`

File: source/reporters/ut_coverage_html_reporter.tps [hunk 4]
```diff
@@ -0,0 +1,60 @@
+create or replace type ut_coverage_html_reporter force under ut_coverage_reporter_base(
```
Comment (Pazus) [inline source/reporters/ut_coverage_html_reporter.tps]: lets not use `force`

File: source/reporters/ut_coverage_xml_reporter.tps [hunk 5]
```diff
@@ -0,0 +1,60 @@
+create or replace type ut_coverage_xml_reporter force under ut_coverage_reporter_base(
```
Comment (Pazus) [inline source/reporters/ut_coverage_xml_reporter.tps]: lets not use `force`

File: docs/userguide/coverage.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
+# Coverage
+utPLSQL comes with build-in coverage reporting engine. The code coverage reporting is based off DBMS_PROFILER package. Code coverage is gathered for the following source types:
+* package bodies
+* type bodies
+* triggers
```
Comment (jgebal) [inline docs/userguide/coverage.md]: will fix

File: docs/userguide/coverage.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+To obtain information about code coverage of your Unit Tests, all you need to do is run your unit tests with one of build-in code coverage reporters.
+Following code coverage reporters are supplied with utPLSQL:
+* `ut_coverage_html_reporter` - generates a HTML coverage report providing summary and detailed information on code coverage. The html reporter is based on open-source [simplecov-html](https://github.com/colszowka/simplecov-html) reporter for Ruby. It includes source code of the code that was covered (if possible)  
+* `ut_coverage_json_reporter` - generates a JSON coverage report providing detailed information on code coverage with line numbers. This coverage report is designed to be consumed by cloud services like [coveralls](https://coveralls.io) 
+* `ut_coverage_xml_reporter`  - generates a JSON coverage report providing detailed information on code coverage with line numbers. This coverage report is designed to be consumed by local services like [sonarqube](https://about.sonarqube.com/)
```
Comment (jgebal) [inline docs/userguide/coverage.md]: will change

File: docs/userguide/coverage.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+Code coverage is using DBMS_PROFILER to gather information about execution of code under test and therefore follows the [DBMS_PROFILER's Security Model](https://docs.oracle.com/database/121/ARPLS/d_profil.htm#ARPLS67465)
+In order to be able to gather coverage information, user executing unit tests needs to be either:
+* Owner of the code that is tested
+* Have the following privileges to be able to gather coverage on code owned by other users:
+    * `create any procedure` system privilege
+    * `execute` privilege on the code that is tested (not only the unit tests) or `execute any procedure` system privilege
+    
+If you have `execute` privilege on the code that are tested, but do not have `create any procedure` system privilege, the code that is tested will be reported as not covered (coverage = 0%).
+If you have `execute` privilege only on the unit tests, but do not have `execute` privilege on the code that is tested, the code will not be reported by coverage - as if it did not exist in the database.
+If the code that is testes is complied as NATIVE, the code coverage will not be reported as well.
```
Comment (jgebal) [inline docs/userguide/coverage.md]: as per Oracle documentation

File: docs/userguide/coverage.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+##Running unite tests with coverage
+Using code coverage functionality is as easy as using any other [reporter](reporters.md) for utPLSQL project. All you need to do is run your tests from your preferred SQL tool and save the outcomes of reporter to a file.
+All you need to do, is pass the constructor of the reporter to your `ut.run` 
+```sql
```
Comment (jgebal) [inline docs/userguide/coverage.md]: ok

File: source/core/coverage/proftab_temp_tables.sql [hunk 10]
```diff
@@ -0,0 +1,233 @@
+create global temporary table plsql_profiler_runs$(
```
Comment (jgebal) [inline source/core/coverage/proftab_temp_tables.sql]: need to rethink it.

File: source/core/coverage/ut_coverage.pkb [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+  * Public functions
+  */
+  function default_file_to_obj_type_map return ut_key_value_pairs is
+  begin
+    return ut_key_value_pairs(
+        ut_key_value_pair('fnc', 'FUNCTION'),
+        ut_key_value_pair('prc', 'PROCEDURE'),
+        ut_key_value_pair('trg', 'TRIGGER'),
+        ut_key_value_pair('tpb', 'TYPE BODY'),
+        ut_key_value_pair('pkb', 'PACKAGE BODY'),
```
Comment (jgebal) [inline source/core/coverage/ut_coverage.pkb]: we don't report on `spec`, added another mapping for `body` extension

File: source/core/coverage/ut_coverage.pkb [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+    a_object_type_subexpression   positive := gc_regex_type_subexpression
+  ) return ut_coverage_file_mappings is
+    type tt_key_values is table of varchar2(4000) index by varchar2(4000);
+    l_key_values tt_key_values;
+    l_mappings   ut_coverage_file_mappings;
+    l_mapping    ut_coverage_file_mapping;
+    l_object_type_key varchar2(4000);
+    l_object_type     varchar2(4000);
+  begin
+    for i in 1 .. a_file_to_object_type_mapping.count loop
```
Comment (jgebal) [inline source/core/coverage/ut_coverage.pkb]: ok

File: source/core/coverage/ut_coverage.pkb [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+    l_skipped_objects  ut_object_names := ut_object_names();
+
+    type t_source_lines is table of binary_integer;
+    l_source_lines     t_source_lines;
+    line_no            binary_integer;
+    l_schema_names     ut_varchar2_list := coalesce(g_schema_names,ut_varchar2_list(sys_context('USERENV','CURRENT_SCHEMA')));
+  begin
+
+    if not ut_coverage_helper.is_develop_mode() then
+      l_skipped_objects := ut_utils.get_utplsql_objects_list() multiset union set(g_skipped_objects);
```
Comment (jgebal) [inline source/core/coverage/ut_coverage.pkb]: changed to coalesce - we don't need union

File: source/core/coverage/ut_coverage_helper.pkb (hunk 1 above)
Comment (jgebal) [inline source/core/coverage/ut_coverage_helper.pkb]: will need to rethink the approach

File: source/reporters/ut_coverage_html_reporter.tps (hunk 4 above)
Comment (jgebal) [inline source/reporters/ut_coverage_html_reporter.tps]: will address it in new and existing files together.

File: docs/userguide/coverage.md [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+        ut_coverage_file_mapping(
+          file_name    => 'sources/hr/betwnstr.fnc',
+          object_owner => 'usr',
+          object_name  => 'betwnstr',
+          object_type  => 'function'                        
+        )
+      )
+    )
+  );
+end;
```
Comment (jgebal) [inline docs/userguide/coverage.md]: I don't understand what you mean.

File: docs/userguide/coverage.md [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+```
+Executes all tests in schema `usr` and reports coverage for that execution on procedure `award_bonus` and function `betwnstr`. The coverage report is mapped-back to file-system object names with paths.
+
+**Reporting using regex file mapping rule**
+If file names and paths in your project follow a well established naming conventions, 
+then you can you can use the predefined rule for mapping file names to object names or you can define your own rule and pass it to the coverage reporter at runtime.
```
Comment (jgebal) [inline docs/userguide/coverage.md]: ok

File: source/core/coverage/proftab_temp_tables.sql (hunk 10 above)
Comment (jgebal) [inline source/core/coverage/proftab_temp_tables.sql]: OK. but then we do not clean up the profiler tables and they will continuously grow and will require clean up/maintenance by DBAs.

File: source/core/coverage/ut_coverage.pkb [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+
+  /**
+  * Public functions
+  */
+  function default_file_to_obj_type_map return ut_key_value_pairs is
+  begin
+    return ut_key_value_pairs(
+        ut_key_value_pair('fnc', 'FUNCTION'),
+        ut_key_value_pair('prc', 'PROCEDURE'),
+        ut_key_value_pair('trg', 'TRIGGER'),
```
Comment (jgebal) [inline source/core/coverage/ut_coverage.pkb]: Can you tell more? What do you thins should be changed?

File: source/core/coverage/ut_coverage.pkb [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+            , a_object_owner, sys_context('USERENV', 'CURRENT_SCHEMA')
+          ),
+          object_name => upper(regexp_substr(a_file_paths(i), a_regex_pattern, 1, 1, 'i', a_object_name_subexpression)),
+          object_type => l_object_type
+        );
+        l_mappings.extend();
+        l_mappings(l_mappings.last) := l_mapping;
+      end loop;
+    end if;
+    return l_mappings;
```
Comment (jgebal) [inline source/core/coverage/ut_coverage.pkb]: good point - taken care of.

File: source/core/coverage/ut_coverage_helper.pkb (hunk 1 above)
Comment (jgebal) [inline source/core/coverage/ut_coverage_helper.pkb]: After a rethink we need a `union`.
If running in `utplsql_develop` mode, we will exclude Unit Test objects from coverage reporting additionally we will exclude all objects that are part of utplsql framework from the coverage.

Added a comment to indicate that.

File: source/core/coverage/ut_coverage_helper.pkb (hunk 2 above)
Comment (jgebal) [inline source/core/coverage/ut_coverage_helper.pkb]: For now, I'll pass the filters to the select.
We will see. If it needs further improvements, we will address this as separate issue.
Processing row by row would require a redesign. We might need to go into solution with pipelined table functions which save memory but are much slower.

Comment (jgebal) [inline source/core/coverage/ut_coverage_helper.pkb]: We're doing it object by object so no issues.

File: docs/userguide/coverage.md (hunk 14 above)
Comment (Pazus) [inline docs/userguide/coverage.md]: Missed it's about stored procedure and function. Please skip this one.

File: source/core/coverage/ut_coverage.pkb (hunk 11 above)
Comment (Pazus) [inline source/core/coverage/ut_coverage.pkb]: Maybe we should still add pkg extension so we at least somewhat cover such clients gathering coverage only for bodies. They won't change the way they develop.

Comment (Pazus) [inline source/core/coverage/ut_coverage.pkb]: In theory is it possible to treat only lines with := in spec while coverage reporting? Or mark all spec lines as skipped so they don't lower coverage but still persist in the report? I think it's more general.

File: source/core/coverage/proftab_temp_tables.sql (hunk 10 above)
Comment (Pazus) [inline source/core/coverage/proftab_temp_tables.sql]: As usual with profiler tables

File: source/core/coverage/ut_coverage.pkb (hunk 16 above)
Comment (Pazus) [inline source/core/coverage/ut_coverage.pkb]: Seems you have already fixed that. Trigger was mapped twice.

File: source/core/coverage/ut_coverage_helper.pkb (hunk 2 above)
Comment (Pazus) [inline source/core/coverage/ut_coverage_helper.pkb]: Ok up to you:)
Personally  I like "for cursor" more in such case as it consumes less memory as it fetches only 255 rows at once:)

**Comment by jgebal** [comment]
All the requested changes have been  done. 
I used rebase on my branch to update from develop to update

**Comment by Pazus** [comment]
Rebase :( Then I merge it

**Pull request merged**